File: tests/fixtures/python/r002_log_with_redaction.py
"""Fixture: should NOT trigger R002.

Each log call passes through a redactor / hasher / length projection
before reaching the log sink, OR logs structured fields the developer
controls (no raw user / LLM content).
"""
import nemoguardrails  # noqa: F401  (suppresses DF001)
import structlog  # noqa: F401  (suppresses R001)

import hashlib
import logging

from flask import Flask, request
from langchain_openai import ChatOpenAI


app = Flask(__name__)
log = logging.getLogger(__name__)
chain = ChatOpenAI(model="gpt-4o-mini")


def redact(s: str) -> str:
    """In-house redactor — strips PII before logging."""
    return s  # implementation elided; pattern-name is the signal


@app.route("/chat-hash", methods=["POST"])
def chat_hash():
    prompt = request.json["q"]
    prompt_hash = hashlib.sha256(prompt.encode()).hexdigest()
    log.info(f"User asked (hash={prompt_hash})")  # OK — hash, not content

    response = chain.invoke(prompt)
    response_hash = hashlib.sha256(response.content.encode()).hexdigest()
    log.info(f"Model returned (hash={response_hash})")  # OK
    return response.content


@app.route("/chat-redact", methods=["POST"])
def chat_redact():
    prompt = request.json["q"]
    log.info(f"User asked: {redact(prompt)}")  # OK — redactor on the path

    response = chain.invoke(prompt)
    log.info(f"Model returned: {redact(response.content)}")  # OK
    return response.content


@app.route("/chat-len", methods=["POST"])
def chat_len():
    prompt = request.json["q"]
    log.info(f"prompt len={len(prompt)}")  # OK — length only, content lost

    response = chain.invoke(prompt)
    log.info("response_len=%d", len(response.content))  # OK
    return response.content


@app.route("/chat-fields", methods=["POST"])
def chat_fields():
    prompt = request.json["q"]
    response = chain.invoke(prompt)
    # Logging structured fields we control — no raw prompt/completion content.
    log.info({"role": "user", "msg_id": "abc-123", "model": "gpt-4o-mini"})
    return response.content

File: tests/fixtures/java/r002_log_with_redaction.java
// Fixture: should NOT trigger R002-java.
// Each log call passes through a redactor / hasher / length projection
// before reaching the log sink.
package com.example.safe;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;

@RestController
public class LogRedacted {

    private static final Logger log = LoggerFactory.getLogger(LogRedacted.class);

    private final ChatClient chatClient;
    private final Redactor redactor;

    public LogRedacted(ChatClient chatClient, Redactor redactor) {
        this.chatClient = chatClient;
        this.redactor = redactor;
    }

    @PostMapping("/chat-hash")
    public String chatHash(@RequestParam String prompt) throws Exception {
        // Hash the prompt — log the hash, not the value.
        byte[] hashBytes = MessageDigest.getInstance("SHA-256")
                .digest(prompt.getBytes(StandardCharsets.UTF_8));
        String promptHash = Base64.getEncoder().encodeToString(hashBytes);
        log.info("User asked (hash={})", promptHash);

        String response = chatClient.prompt().user(prompt).call().content();
        byte[] respBytes = MessageDigest.getInstance("SHA-256")
                .digest(response.getBytes(StandardCharsets.UTF_8));
        log.info("Model returned (hash={})", Base64.getEncoder().encodeToString(respBytes));
        return response;
    }

    @PostMapping("/chat-redact")
    public String chatRedact(@RequestParam String prompt) {
        log.info("User asked: {}", redactor.redact(prompt));

        String response = chatClient.prompt().user(prompt).call().content();
        log.info("Model returned: {}", redactor.redact(response));
        return response;
    }

    @PostMapping("/chat-len")
    public String chatLen(@RequestParam String prompt) {
        log.info("prompt len={}", prompt.length());

        String response = chatClient.prompt().user(prompt).call().content();
        log.info("response len={}", response.length());
        return response;
    }

    public interface Redactor {
        String redact(String s);
    }
}
